File: leetcode/medium/SortAnArray.py
import random
from typing import List

class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        """
        Sorts an array of integers in ascending order using the Mergesort algorithm.

        This implementation uses a recursive, top-down Mergesort. It divides the
        array into two halves, sorts them recursively, and then merges them back
        together. This guarantees a time complexity of O(n log n).

        Args:
            nums: The list of integers to sort.

        Returns:
            The sorted list of integers.

        Time Complexity: O(n log n) for all cases (best, average, worst).
        Space Complexity: O(n) due to the temporary array used in the merge step.
        """
        self.mergesort(nums, 0, len(nums) - 1)
        return nums
    
    def mergesort(self, nums: List[int], low: int, high: int):
        """
        Recursively sorts a sub-array `nums[low..high]` using the Mergesort algorithm.
        """
        # Base case: If the sub-array has 0 or 1 elements, it's already sorted.
        if low < high:
            # Find the middle point to divide the array into two halves.
            mid = (low + high) // 2
            
            # Recursively sort the first half.
            self.mergesort(nums, low, mid)
            # Recursively sort the second half.
            self.mergesort(nums, mid + 1, high)
            
            # Merge the two sorted halves.
            self.merge(nums, low, mid, high)

    def merge(self, nums: List[int], low: int, mid: int, high: int):
        """
        Merges two sorted sub-arrays: nums[low..mid] and nums[mid+1..high].
        """
        # Create a temporary array to store the merged result.
        temp = [0] * (high - low + 1)
        
        # Initialize pointers for the two sub-arrays and the temporary array.
        i, j, k = low, mid + 1, 0

        # Compare elements from both sub-arrays and copy the smaller one to temp.
        while i <= mid and j <= high:
            if nums[i] <= nums[j]:
                temp[k] = nums[i]
                i += 1
            else:
                temp[k] = nums[j]
                j += 1
            k += 1
        
        # If there are any remaining elements in the first sub-array, copy them.
        while i <= mid:
            temp[k] = nums[i]
            i += 1
            k += 1
            
        # If there are any remaining elements in the second sub-array, copy them.
        while j <= high:
            temp[k] = nums[j]
            j += 1
            k += 1
            
        # Copy the sorted elements from the temporary array back into the original array.
        for i in range(low, high + 1):
            nums[i] = temp[i - low]

File: leetcode/medium/SortAnArray.java
/**
 * Implements the Heap Sort algorithm to sort an array of integers in ascending order.
 * This version uses an optimized, non-recursive (iterative) heapify method.
 * Time Complexity: O(n log n).
 * Space Complexity: O(1) (in-place sorting).
 */
class Solution {
    
    /**
     * Sorts the given array of integers using the Heap Sort algorithm.
     *
     * @param nums The array to be sorted.
     * @return The sorted array.
     */
    public int[] sortArray(int[] nums) {
        if (nums == null || nums.length < 2) {
            return nums;
        }

        int length = nums.length;
        
        // 1. Build Max-Heap: Rearrange array into a Max-Heap.
        // Start from the last non-leaf node (length/2 - 1) up to the root (0).
        // 
        for (int i = length / 2 - 1; i >= 0; i--) {
            // The 0-based index of the last non-leaf node is floor(length/2) - 1.
            heapify(nums, i, length);
        }

        // 2. Heap Sort: Extract elements one by one from the heap.
        // The array will be sorted in place.
        for (int i = length - 1; i > 0; i--) {
            // Swap the current root (largest element) with the last element of the unsorted part (index i).
            swap(nums, 0, i);

            // Call max-heapify on the reduced heap (size 'i').
            // The largest element is now at index i, which is excluded from the new heap.
            heapify(nums, 0, i);
        }
        
        return nums;
    }

    // --- Core Helper Functions ---

    /**
     * Iteratively maintains the Max-Heap property in a subtree rooted at 'start'.
     * This function is non-recursive.
     *
     * @param nums The array representing the heap.
     * @param start The index of the root of the subtree to heapify.
     * @param end The exclusive boundary of the heap (the size of the heap is 'end').
     */
    private void heapify(int[] nums, int start, int end) {
        // 'curr' tracks the current node we are comparing/swapping down the heap.
        int curr = start; 

        // Loop until 'curr' is a leaf node or the Max-Heap property is satisfied at 'curr'.
        while (true) {
            int parent = curr;
            int left = curr * 2 + 1; // Left child index
            int right = curr * 2 + 2; // Right child index
            
            // Assume the current parent is the largest.
            int largest = parent; 

            // 1. Check if the left child exists and is larger than the current largest.
            if (left < end && nums[left] > nums[largest]) {
                largest = left;
            }

            // 2. Check if the right child exists and is larger than the current largest.
            // Note: This correctly compares the right child against the largest found so far (parent or left).
            if (right < end && nums[right] > nums[largest]) {
                largest = right;
            }

            // If the largest element is the parent itself, the heap property is satisfied.
            if (largest == parent) {
                break;
            }

            // If the largest is a child, swap it with the parent.
            swap(nums, parent, largest);
            
            // Move 'curr' down to the child's position to continue heapifying the sub-tree.
            curr = largest; 
        }
    }

    /**
     * Helper method to swap two elements in the array.
     * @param nums The array.
     * @param i Index of the first element.
     * @param j Index of the second element.
     */
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}